resilience4j-ratelimiter/src/main/java/io/github/resilience4j/ratelimiter/internal/RefillRateLimiter.java

```diff
@@ -0,0 +1,476 @@
+/*
+ *
+ *  Copyright 2020 Emmanouil Gkatziouras
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *         http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ *
+ *
+ */
+package io.github.resilience4j.ratelimiter.internal;
+
+import io.github.resilience4j.ratelimiter.RateLimiter;
+import io.github.resilience4j.ratelimiter.RefillRateLimiterConfig;
+import io.github.resilience4j.ratelimiter.event.RateLimiterOnFailureEvent;
+import io.github.resilience4j.ratelimiter.event.RateLimiterOnSuccessEvent;
+import io.vavr.collection.HashMap;
+import io.vavr.collection.Map;
+
+import java.time.Duration;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.function.UnaryOperator;
+
+import static java.lang.Long.min;
+import static java.lang.System.nanoTime;
+import static java.lang.Thread.currentThread;
+import static java.util.concurrent.locks.LockSupport.parkNanos;
+
+/**
+ * {@link RefillRateLimiter} has a max capacity of permits refilled periodically.
+ * <p>Each period has duration of {@link RefillRateLimiterConfig#limitRefreshPeriod} in nanoseconds.
+ * <p>By contract the initial {@link RefillRateLimiter.State#activePermissions} are set
+ * to be the same with{@link RefillRateLimiterConfig#permitCapacity}
+ * <p>A ratio of permit per nanoseconds is calculated using
+ * {@link RefillRateLimiterConfig#limitRefreshPeriod} and {@link RefillRateLimiterConfig#limitForPeriod}.
+ * On a permit request the permits that should have been replenished are calculated based on the nanos passed and the ratio of nanos per permit.
+ * For the {@link RefillRateLimiter} callers it is looks like a token bucket supporting bursts and a gradual refill,
+ * under the hood there is some optimisations that will skip this refresh if {@link RefillRateLimiter} is not used actively.
+ * <p>All {@link RefillRateLimiter} updates are atomic and state is encapsulated in {@link
+ * AtomicReference} to {@link RefillRateLimiter.State}
+ */
+public class RefillRateLimiter implements RateLimiter {
```

Comment:
I fill like the majority of this code is the same as in AtomicRateLimiter. Maybe you can inherit from it and override only changed methods.